### ISSUE TITLE
Live migration fails with permission error on /var/run/ganeti/kvm-hypervisor

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1231]
Originally reported of Google Code with ID 1178.

```
*What software version are you running?*

root@nuc2:~# gnt-cluster --version
gnt-cluster (ganeti 2.15.2-3) 2.15.2

root@nuc2:~# gnt-cluster version
Software version: 2.15.2
Internode protocol: 2150000
Configuration format: 2150000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version 2.15.2-3

root@nuc2:~# hspace --version
hspace (ganeti) version 2.15.2-3
compiled with ghc 7.10
running on linux x86_64

*What distribution are you using?*

This is a 2-node test cluster. nuc1 is Debian Jessie (with 2.15.2 from jessie-backports), nuc2 is Ubuntu 16.04 (also has 2.15.2). Using kvm as the hypervisor.

*What steps will reproduce the problem?*

Trying to live-migrate a drbd instance from one node to the other.

...
Sun May  8 13:08:18 2016 Migrating instance windows0
Sun May  8 13:08:19 2016 * warning: hypervisor version mismatch between source ([2, 1, 2]) and target ([2, 5, 0]) node
Unhandled Ganeti error: Cannot change directory permissions on '/var/run/ganeti/kvm-hypervisor': [Errno 1] Operation not permitted: '/var/run/ganeti/kvm-hypervisor'

I made a small patch to find out what mode it was trying to set:

--- utils/io.py.orig	2016-03-26 07:28:53.000000000 +0000
+++ utils/io.py	2016-05-08 13:07:53.991432226 +0100
@@ -605,7 +605,7 @@
       os.chmod(dir_name, dir_mode)
     except EnvironmentError, err:
       raise errors.GenericError("Cannot change directory permissions on"
-                                " '%s': %s" % (dir_name, err))
+                                " '%s' to 0%o: %s" % (dir_name, dir_mode, err))
     if not os.path.isdir(dir_name):
       raise errors.GenericError("%s is not a directory" % dir_name)


Now running again, with -d flag as well:

root@nuc2:~# gnt-instance migrate -d windows0
2016-05-08 13:15:28,371: gnt-instance migrate pid=14395 cli:1218 DEBUG Command line: gnt-instance migrate -d windows0
Instance windows0 will be migrated. Note that migration might impact
the instance if anything goes wrong (e.g. due to bugs in the
hypervisor). Continue?
y/[n]/?: y
Sun May  8 13:15:33 2016 Migrating instance windows0
Sun May  8 13:15:33 2016 * warning: hypervisor version mismatch between source ([2, 1, 2]) and target ([2, 5, 0]) node
2016-05-08 13:15:33,869: gnt-instance migrate pid=14395 cli:1225 ERROR Error during command processing
Traceback (most recent call last):
  File "/usr/share/ganeti/2.15/ganeti/cli.py", line 1221, in GenericMain
    result = func(options, args)
  File "/usr/share/ganeti/2.15/ganeti/client/gnt_instance.py", line 820, in MigrateInstance
    SubmitOrSend(op, cl=cl, opts=opts)
  File "/usr/share/ganeti/2.15/ganeti/cli.py", line 1011, in SubmitOrSend
    return SubmitOpCode(op, cl=cl, feedback_fn=feedback_fn, opts=opts)
  File "/usr/share/ganeti/2.15/ganeti/cli.py", line 976, in SubmitOpCode
    reporter=reporter)
  File "/usr/share/ganeti/2.15/ganeti/cli.py", line 955, in PollJob
    return GenericPollJob(job_id, _LuxiJobPollCb(cl), reporter)
  File "/usr/share/ganeti/2.15/ganeti/cli.py", line 777, in GenericPollJob
    errors.MaybeRaise(msg)
  File "/usr/share/ganeti/2.15/ganeti/errors.py", line 531, in MaybeRaise
    raise errcls(*args)
GenericError: Cannot change directory permissions on '/var/run/ganeti/kvm-hypervisor' to 0775: [Errno 1] Operation not permitted: '/var/run/ganeti/kvm-hypervisor'
Unhandled Ganeti error: Cannot change directory permissions on '/var/run/ganeti/kvm-hypervisor' to 0775: [Errno 1] Operation not permitted: '/var/run/ganeti/kvm-hypervisor'

(Aside: the backtrace shown appears to be in the code which submits the job - not the code which executes the job. It would be nice if it serialized the backtrace too)

The actual permissions on both nodes are the same and already with the correct mode:

root@nuc1:~# ls -ld /var/run/ganeti/kvm-hypervisor/
drwxrwxr-x 10 root root 200 May  8 09:25 /var/run/ganeti/kvm-hypervisor/

root@nuc2:~# ls -ld /var/run/ganeti/kvm-hypervisor/
drwxrwxr-x 10 root root 200 May  8 09:32 /var/run/ganeti/kvm-hypervisor/

I found the same error raised in issue #608 but this appears to be unrelated (it's to do with console, not live migration)

It appears that the code is running as another user (gnt-masterd?) which doesn't have rights to chmod.

So I made a simple workaround, which avoids doing chmod unless the permissions are not already correct:

--- /usr/share/ganeti/2.15/ganeti/utils/io.py.orig	2016-03-26 07:28:53.000000000 +0000
+++ /usr/share/ganeti/2.15/ganeti/utils/io.py	2016-05-08 13:26:33.267923013 +0100
@@ -601,11 +601,12 @@
       if err.errno != errno.EEXIST:
         raise errors.GenericError("Cannot create needed directory"
                                   " '%s': %s" % (dir_name, err))
-    try:
-      os.chmod(dir_name, dir_mode)
-    except EnvironmentError, err:
-      raise errors.GenericError("Cannot change directory permissions on"
-                                " '%s': %s" % (dir_name, err))
+    if (os.stat(dir_name).st_mode & 07777) != dir_mode:
+      try:
+        os.chmod(dir_name, dir_mode)
+      except EnvironmentError, err:
+        raise errors.GenericError("Cannot change directory permissions on"
+                                  " '%s' to 0%o: %s" % (dir_name, dir_mode, err))
     if not os.path.isdir(dir_name):
       raise errors.GenericError("%s is not a directory" % dir_name)

After this, live migration works.

(Aside: it actually works in the direction from older Jessie kernel/KVM to newer Xenial kernel/KVM. In the other direction ganeti *thinks* it migrated successfully, but actually it goes straight into ERROR_down state. But that's not relevant to the issue in hand)
```
Originally added on 2016-05-08 12:48:45 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

clearly EnsureDir is doing more work than it should be, and we should fix it, but there's something a bit strange here.

The migrate is actually performed by RPC calls to noded. It sends accept_instance to the target and instance_migrate to the source. By default noded should be running as the default user, which is root. This can be overridden with the --with-default-user ./configure option.

Could you check what user your noded is running as? For me
# grep  '_USER =' /etc/ganeti/share/ganeti/_constants.py

MASTERD_USER = "gnt-masterd"
METAD_USER = "gnt-metad"
RAPI_USER = "gnt-rapi"
CONFD_USER = "gnt-confd"
WCONFD_USER = "gnt-masterd"
KVMD_USER = "root"
LUXID_USER = "gnt-masterd"
NODED_USER = "root"
MOND_USER = "root"
SSH_LOGIN_USER = "root"
SSH_CONSOLE_USER = "root"
HT_SM_USER = "user"
OPCODE_REASON_SRC_USER = "gnt:user"

```
Originally added on 2016-05-09 10:41:24 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Only difference is that gnt-metad is running as root:

root@nuc1:~# grep '_USER =' /etc/ganeti/share/ganeti/_constants.py
MASTERD_USER = "gnt-masterd"
METAD_USER = "root"
RAPI_USER = "gnt-rapi"
CONFD_USER = "gnt-confd"
WCONFD_USER = "gnt-masterd"
KVMD_USER = "root"
LUXID_USER = "gnt-masterd"
NODED_USER = "root"
MOND_USER = "root"
SSH_LOGIN_USER = "root"
SSH_CONSOLE_USER = "root"
HT_SM_USER = "user"
OPCODE_REASON_SRC_USER = "gnt:user"

However, I have now finished upgrading the cluster to a two-node Ubuntu Xenial: it was mixed Jessie+Xenial when I found the problem.

But now I put io.py back to how it was, it's all working fine :-(

Presumably it's either to do with the jessie-backports 2.15.2 package, or something to do with the mixed cluster (e.g. mixed drbd or kvm versions)

Anyway I'm happy if you want to just close this.
```
Originally added on 2016-05-10 18:30:43 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Brian (and Brian) - I'm seeing the same issue here on a 2.15 cluster with a similar config. as Brian's: 2.15.2 on both nodes, one Jessie, one running 16.04 LTS.

Looking at /etc/ganeti/share/ganeti/_constants.py:

- on the jessie node:

MASTERD_USER = "gnt-masterd"
METAD_USER = "root"
RAPI_USER = "gnt-rapi"
CONFD_USER = "gnt-confd"
WCONFD_USER = "gnt-masterd"
KVMD_USER = "root"
LUXID_USER = "gnt-masterd"
NODED_USER = "root"
MOND_USER = "root"
SSH_LOGIN_USER = "root"
SSH_CONSOLE_USER = "root"
HT_SM_USER = "user"
OPCODE_REASON_SRC_USER = "gnt:user"

- on the 16.04 node:

MASTERD_USER = "gnt-masterd"
METAD_USER = "root"
RAPI_USER = "gnt-rapi"
CONFD_USER = "gnt-confd"
WCONFD_USER = "gnt-masterd"
KVMD_USER = "root"
LUXID_USER = "gnt-masterd"
NODED_USER = "root"
MOND_USER = "root"
SSH_LOGIN_USER = "root"
SSH_CONSOLE_USER = "root"
HT_SM_USER = "user"
OPCODE_REASON_SRC_USER = "gnt:user"

... I can easily reproduce this if it helps finding the cause. 
```
Originally added on 2016-09-26 20:25:23 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I am seeing this problem again, also in a cluster mid-way through upgrade.

root@wrn-vm2:~# gnt-instance migrate wrn-smb1
Instance wrn-smb1 will be migrated. Note that migration might impact
the instance if anything goes wrong (e.g. due to bugs in the
hypervisor). Continue?
y/[n]/?: y
Thu Nov 10 13:13:55 2016 Migrating instance wrn-smb1.int.example.net
Thu Nov 10 13:13:55 2016 * warning: hypervisor version mismatch between source ([2, 1, 2]) and target ([2, 5, 0]) node
Unhandled Ganeti error: Cannot change directory permissions on '/var/run/ganeti/kvm-hypervisor': [Errno 1] Operation not permitted: '/var/run/ganeti/kvm-hypervisor'

vm2 (old node) is running debian jessie + 2.15.2 from backports, and vm1 (new node) is running ubuntu xenial + 2.15.2. The instance to be migrated is on vm2.

On both nodes, /var/run/ganeti/kvm-hypervisor is mode 775 and owned by root:root.

On both nodes, noded is running as root.

root@wrn-vm1:~# ps auxwww | grep noded
root      3493  0.0  0.2 154148 39612 ?        SL   12:50   0:00 /usr/bin/python /usr/sbin/ganeti-noded
root      6420  0.0  0.0  14228   928 pts/0    S+   13:29   0:00 grep --color=auto noded

I made the patch on wrn-vm1 (new) first, but that didn't fix the problem; it only went away after I patched wrn-vm2 as well.

```
Originally added on 2016-11-10 13:36:55 +0000 UTC.

**Comment by StricUmu** [comment]
Ah, this had a nice trick up its sleeves, and only happens when hypervisor versions differ.

Problem as follows:
* /var/run/ganeti/kvm-hypervisor is owned by root:root, and this seems fine in normal operations
* in cmdlib/instance_migration.py in TLMigrateInstance::_ExecMigration
** if there's a different hypervisor version then it runs hypervisor.GetHypervisor(self.instance.hypervisor) to find an object that can figure out if src_version and dst_version are friends.
** This seems to make a call to gnt-masterd, which is run as gnt-masterd, which through some path comes into utils/io.py in EnsureDirs and tries to chown the root owned /var/run/ganeti/kvm-hypervisor (while being user gnt-masterd). This fails.

Crappy workaround on master: chown -R gnt-masterd /var/lib/ganeti/kvm-hypervisor

Potential fix: As patch earlier, don't just blindly chmod - but rather check if anything needs to be done (what if it does? then we're still running as the wrong user)

Actual fix: Someone who knows what's supposed to happen do something :)

**Comment by xals** [comment]
Hi there, I just had this buggy behaviour because we have installed the first Debian Stretch node in our cluster (the other nodes run Debian Jessie), and the KVM version is not the same (2.1 on Jessie nodes and 2.8 on the Stretch one). And I forgot to upgrade qemu-kvm to the jessie-backports version (2.8).

The fix worked for me, but the directory is `/var/run/ganeti/kvm-hypervisor`, but not `/var/lib`.

Thanks!

**Comment by zitterbacke** [comment]
+1 while upgrading from SLES12 SP1 to SP2 (with ganeti from git heads/stable-2.14)

**Comment by saschalucas** [comment]
Hi,

any news on this? My personal opinion is, that the proposed fix in utils/io.py should be accepted. This wouldn't interrupt users doing rolling node upgrades.

As I understand it, the blocker is, that no one can explain, why this is called? Maybe this can be investigated afterwards?

Thanks, Sascha.

**Comment by iustin** [comment]
Sent a simplified version of the code proposed. With logs, it could be possible to hunt down the actual root cause.

**Comment by iustin** [comment]
OK, while the mitigation is in place, the question is why this happens in the first place.

The comment from @StricUmu seems quite to the point - yes, masterd behaves differently when the versions are mismatched, but:

- `_ExecMigration` calls GetHypervisorClass() - which doesn't instantiate the hypervisor, only returns the class.
- and then it calls `VersionSafeForMigration`, which is indeed a staticmethod.
- furthermore, the KVM hypervisor calls `EnsureDirs` for `/var/run/kvm-hypervisor` only in the `__init__`, so should not happen when just using the class.
- a quick grep in all of cmdlib code doesn't show any calls to `GetHypervisor`, only `GetHypervisorClass`

So based on current stable-2.15, this shouldn't happen. And there's no related diff from 2.14 to 2.15 for `instance_migration.py` nor kvm hypervisor code.

I guess only actual live testing will explain how this happens… Somebody, somewhere breaks the rule that hypervisors shouldn't be instantiated in masterd code.

**Comment by saschalucas** [comment]
Hi Iustin, thanks for working on this issue. I'll try to produce the trace you need. But I'm unsure how to get it. You are talking about masterd. Is this luxid, wconfd oder noded on master? The only option that comes to my mind is to run a daemon in debug-mode (-d). THX, Sascha.

**Comment by iustin** [comment]
Ee, I'm probably using very old terminology. Not sure if there's a masterd anymore :)

Basically there is python code executed on the master - not in noded - that does the wrong thing. I should probably understand exactly what the 2.15/2.16 architecture is…

**Comment by saschalucas** [comment]
So the error message is found in python code. The submission of the migrate job shows no traces how this code is reached. So it is assumed, that the job execution is the cause. AFAIK luxid and wconfd are written in haskell and noded in python (and yes, there is no masterd anymore). Both (luxid and wconfd) run as gnt-masterd. Right the user from which the error seems to come from. So there must be something in haskell calling python directly (and just in case of hypervisor version mismatch)?

**Comment by iustin** [comment]
Ooh, interesting architecture. Will definitely have to brush up on my knowledge :) Let my try to get some time to bring up a ganeti mini-cluster and see how it works with modern versions, and then I'll update this thread.

**Comment by saschalucas** [comment]
This is fixed with #1281 and working with Ganeti 2.16.2 as tested by my self with upgrade from Ubuntu Xenial to Bionic.
